### ISSUE TITLE
curl error - triton deployment inference

### ISSUE THREAD

**Issue by HalteroXHunter** [triton-inference-server/tensorrtllm_backend#335]
Once I have correcly deployed my model on a triton server, once I try to send a request:

`curl -X POST localhost:8000/v2/models/ensemble/generate -d '{"text_input": "What is machine learning?", "max_tokens": 100, "bad_words": "", "stop_words": ""}'`

I get the following error:

`{"error":"in ensemble 'ensemble', [request id: <id_unknown>] unexpected deadlock, at least one output is not set while no more ensemble steps can be made"}`

any idea what is going on?

Including the config files

also for the build.py these are the used params:

`python build.py --model_dir ./c-model/gpt2/1-gpu/ \
                --dtype float16 \
                --use_inflight_batching \
                --remove_input_padding \
                --use_gpt_attention_plugin float16 \
                --paged_kv_cache \
                --enable_context_fmha \
                --use_gemm_plugin float16 \
                --use_parallel_embedding \
                --max_input_len=1024 \
                --max_output_len=1024 \
                --max_batch_size 32 \
                --world_size 1 \
                --output_dir engines/fp16/1-gpu`

[ensemble_config.txt](https://github.com/triton-inference-server/tensorrtllm_backend/files/14312964/ensemble_config.txt)
[tensor_llm_bls_config.txt](https://github.com/triton-inference-server/tensorrtllm_backend/files/14312965/tensor_llm_bls_config.txt)
[tensorllm_config.txt](https://github.com/triton-inference-server/tensorrtllm_backend/files/14312966/tensorllm_config.txt)
[preprocessing_config.txt](https://github.com/triton-inference-server/tensorrtllm_backend/files/14312967/preprocessing_config.txt)
[postprocessing_config.txt](https://github.com/triton-inference-server/tensorrtllm_backend/files/14312968/postprocessing_config.txt)

**Comment by HalteroXHunter** [comment]
@kaiyux please help

**Comment by byshiue** [comment]
Please share the full reproduced steps, including how do you build docker image, how do you build engine, launching serving and send request.

**Comment by HalteroXHunter** [comment]
> Please share the full reproduced steps, including how do you build docker image, how do you build engine, launching serving and send request.

Hello @byshiue :)

What I want to actually do is, at inference, being able to return the logits which are necessary for some tasks in the following repo: https://github.com/EleutherAI/lm-evaluation-harness

I tried to do so using  the following https://github.com/triton-inference-server/tensorrtllm_backend#:~:text=Return%20context%20logits,).%20For%20example%3A however with the current description or maybe my setup, I am not able to execute it.

regarding the following you asked for, these are the following commands:

Docker image build with 23.12 version

# BUILD IMAGE
`DOCKER_BUILDKIT=1 docker build -t triton_trt_llm -f dockerfile/Dockerfile.trt_llm_backend .` 

# DEPLOY FROM CONTAINER
`docker run --rm -it --gpus='"device=2"' --network host --ulimit memlock=-1 --ulimit stack=134217728 --shm-size=4g -v $(pwd)/triton-repository:/tensorrtllm_backend triton_trt_llm bash`

# Go to the tensorrt_llm/examples/gpt directory
	cd tensorrt_llm/examples/gpt

# Download weights from HuggingFace Transformers
rm -rf gpt2 && git clone https://huggingface.co/gpt2-medium gpt2
pushd gpt2 && rm pytorch_model.bin model.safetensors && wget -q https://huggingface.co/gpt2-medium/resolve/main/pytorch_model.bin && popd

# Convert weights from HF Tranformers to FT format
python3 hf_gpt_convert.py -p 8 -i gpt2 -o ./c-model/gpt2 --tensor-parallelism 1 --storage-type float16

python build.py --model_dir ./c-model/gpt2/1-gpu/ \
                --dtype float16 \
                --use_inflight_batching \
                --remove_input_padding \
                --use_gpt_attention_plugin float16 \
                --paged_kv_cache \
                --enable_context_fmha \
                --use_gemm_plugin float16 \
                --use_parallel_embedding \
                --max_input_len=1024 \
                --max_output_len=1024 \
                --max_batch_size 32 \
                --world_size 1 \
                --output_dir engines/fp16/1-gpu

# Copy the TRT engine to triton_model_repo/tensorrt_llm/1/
cp /app/tensorrt_llm/examples/gpt/engines/fp16/1-gpu/* /tensorrtllm_backend/triton_model_repo/tensorrt_llm/1
	
# launch triton server
python3 /app/scripts/launch_triton_server.py --world_size=1 --model_repo=/tensorrtllm_backend/triton_model_repo/

**Comment by pcastonguay** [comment]
hi @HalteroXHunter, would it be possible for you to try with 24.01 branch? Also could you launch the triton server with the `--log` option and share the `triton_log.txt` log file?

```
python3 /app/scripts/launch_triton_server.py --log --world_size=1 --model_repo=/tensorrtllm_backend/triton_model_repo/
```

Thanks

**Comment by pcastonguay** [comment]
To get logits back from TRT-LLM, you would need to build your engine with `--gather_all_token_logits` option. See `Optional outputs` section here: https://github.com/NVIDIA/TensorRT-LLM/blob/main/docs/source/gpt_runtime.md

If using the Triton backend, you would also need to set those input tensors to true: https://github.com/triton-inference-server/tensorrtllm_backend/blob/main/all_models/inflight_batcher_llm/tensorrt_llm/config.pbtxt#L177-L190. You can use one of the provided client for that. See https://github.com/triton-inference-server/tensorrtllm_backend/tree/main?tab=readme-ov-file#return-context-logits-andor-generation-logits

Hope this helps.

**Comment by HalteroXHunter** [comment]
> To get logits back from TRT-LLM, you would need to build your engine with `--gather_all_token_logits` option. See `Optional outputs` section here: https://github.com/NVIDIA/TensorRT-LLM/blob/main/docs/source/gpt_runtime.md
> 
> If using the Triton backend, you would also need to set those input tensors to true: https://github.com/triton-inference-server/tensorrtllm_backend/blob/main/all_models/inflight_batcher_llm/tensorrt_llm/config.pbtxt#L177-L190. You can use one of the provided client for that. See https://github.com/triton-inference-server/tensorrtllm_backend/tree/main?tab=readme-ov-file#return-context-logits-andor-generation-logits
> 
> Hope this helps.

Thanks very much for the help, however I am running into another error on the tensorllm when config when deploying. these are the logs:

![image](https://github.com/triton-inference-server/tensorrtllm_backend/assets/97960232/66e5e687-86be-40b4-be3c-1369ee09271a)
![image](https://github.com/triton-inference-server/tensorrtllm_backend/assets/97960232/d9a3c578-fb8a-4e49-b5e4-25c0980a3659)
![image](https://github.com/triton-inference-server/tensorrtllm_backend/assets/97960232/b1b0b555-1524-4a23-af52-ca5571b53704)
![image](https://github.com/triton-inference-server/tensorrtllm_backend/assets/97960232/52f87a86-0437-4657-bfc4-19649c63e4b6)


These are the steps I follow:

>> initialize the container

docker run --rm -it --gpus='"device=2"' --network host --ulimit memlock=-1 --ulimit stack=134217728 --shm-size=4g -v $(pwd)/triton-repository:/tensorrtllm_backend triton_trt_llm_2401 bash

	once there:

	>Go to the tensorrt_llm/examples/gpt directory
	cd tensorrt_llm/examples/gpt

	>Download weights from HuggingFace Transformers
	rm -rf gpt2 && git clone https://huggingface.co/gpt2-medium gpt2
	pushd gpt2 && rm pytorch_model.bin model.safetensors && wget -q https://huggingface.co/gpt2-medium/resolve/main/pytorch_model.bin && popd

	>Convert weights from HF Tranformers to FT format
	python3 hf_gpt_convert.py -p 8 -i gpt2 -o ./c-model/gpt2 --tensor-parallelism 1 --storage-type float16

	python build.py --model_dir ./c-model/gpt2/1-gpu/ \
                --dtype float16 \
                --use_inflight_batching \
                --gather_all_token_logits \
                --remove_input_padding \
                --use_gpt_attention_plugin float16 \
                --paged_kv_cache \
                --enable_context_fmha \
                --use_gemm_plugin float16 \
                --use_parallel_embedding \
                --max_input_len=1024 \
                --max_output_len=1024 \
                --max_batch_size 32 \
                --world_size 1 \
                --output_dir engines/fp16/1-gpu


    >Copy the TRT engine to triton_model_repo/tensorrt_llm/1/
	cp /app/tensorrt_llm/examples/gpt/engines/fp16/1-gpu/* /tensorrtllm_backend/triton_model_repo/tensorrt_llm/1
	
	>launch triton server
	python3 /app/scripts/launch_triton_server.py --world_size=1 --model_repo=/tensorrtllm_backend/triton_model_repo/


Also find attached the config file for tensorllm
[config.txt](https://github.com/triton-inference-server/tensorrtllm_backend/files/14600046/config.txt)


Thanks in advance,

Ian

**Comment by pcastonguay** [comment]
Based on the error you shared, the `deviceId` specified in your config.pbtxt is incorrect. What's the output of `nvidia-smi` inside the container? I'm assuming the deviceId is 0 inside the container so you should set `gpu_device_ids` to 0 in your config.pbtxt.

**Comment by HalteroXHunter** [comment]
> Based on the error you shared, the `deviceId` specified in your config.pbtxt is incorrect. What's the output of `nvidia-smi` inside the container? I'm assuming the deviceId is 0 inside the container so you should set `gpu_device_ids` to 0 in your config.pbtxt.

Hello and thanks for the response! I changed the deviceId and got it running, however now the logits are showing some resistance.
These are the results and issues I got:

config:

Branch: 24.01

`docker run --rm -it --gpus='"device=2"' --network host --ulimit memlock=-1 --ulimit stack=134217728 --shm-size=4g -v $(pwd)/triton-repository:/tensorrtllm_backend triton_trt_llm_2401 bash

	once there:

	# Go to the tensorrt_llm/examples/gpt directory
	cd tensorrt_llm/examples/gpt

	# Download weights from HuggingFace Transformers
	rm -rf gpt2 && git clone https://huggingface.co/gpt2-medium gpt2
	pushd gpt2 && rm pytorch_model.bin model.safetensors && wget -q https://huggingface.co/gpt2-medium/resolve/main/pytorch_model.bin && popd

	# Convert weights from HF Tranformers to FT format
	python3 hf_gpt_convert.py -p 8 -i gpt2 -o ./c-model/gpt2 --tensor-parallelism 1 --storage-type float16

	python build.py --model_dir ./c-model/gpt2/1-gpu/ \
                --dtype float16 \
                --use_inflight_batching \
                --gather_all_token_logits \
                --remove_input_padding \
                --use_gpt_attention_plugin float16 \
                --paged_kv_cache \
                --enable_context_fmha \
                --use_gemm_plugin float16 \
                --use_parallel_embedding \
                --max_input_len=1024 \
                --max_output_len=1024 \
                --max_batch_size 32 \
                --world_size 1 \
                --output_dir engines/fp16/1-gpu


    # Copy the TRT engine to triton_model_repo/tensorrt_llm/1/
	cp /app/tensorrt_llm/examples/gpt/engines/fp16/1-gpu/* /tensorrtllm_backend/triton_model_repo/tensorrt_llm/1
	
	# launch triton server
	python3 /app/scripts/launch_triton_server.py --world_size=1 --model_repo=/tensorrtllm_backend/triton_model_repo/

if I set:

parameters: {
  key: "gpt_model_type"
  value: {
    string_value: "inflight_fused_batching"
  }
}

I get the following error in deployment:

![image](https://github.com/triton-inference-server/tensorrtllm_backend/assets/97960232/795f7176-eac7-4ddb-a3b9-75fe676169c5)

If I change it to V1 it works, but when I run the inference below I get the following error

# Inference
curl -X POST localhost:8000/v2/models/ensemble/generate -d '{"text_input": "What is machine learning?", "max_tokens": 20, "bad_words": "", "stop_words": "","return-context-logits": true, "return-generation-logits": true}'`


#error
![image](https://github.com/triton-inference-server/tensorrtllm_backend/assets/97960232/44d30329-188f-4edc-b79b-feb9af753ed8)

config for tensorllm: 
[config.txt](https://github.com/triton-inference-server/tensorrtllm_backend/files/14617146/config.txt)

config for ensemble:
[config_ensemble.txt](https://github.com/triton-inference-server/tensorrtllm_backend/files/14617152/config_ensemble.txt)


Thanks in advance!

Regards,

Ian
